app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/room/viewmodel/VideosViewModel.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package android.support.v17.leanback.supportleanbackshowcase.app.room.viewmodel;
+
+import android.app.Application;
+import android.arch.core.util.Function;
+import android.arch.lifecycle.AndroidViewModel;
+import android.arch.lifecycle.LiveData;
+import android.arch.lifecycle.MediatorLiveData;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.lifecycle.Observer;
+import android.arch.lifecycle.Transformations;
+import android.support.annotation.Nullable;
+import android.support.v17.leanback.supportleanbackshowcase.app.room.config.AppConfiguration;
+import android.support.v17.leanback.supportleanbackshowcase.app.room.db.DatabaseHelper;
+import android.support.v17.leanback.supportleanbackshowcase.app.room.db.entity.CategoryEntity;
+import android.support.v17.leanback.supportleanbackshowcase.app.room.db.entity.VideoEntity;
+
+import java.util.List;
+
+public class VideosViewModel extends AndroidViewModel {
+    // live data connect to database
+    private LiveData<List<CategoryEntity>> mAllCategories;
+    private LiveData<List<VideoEntity>> mSearchResults;
+    private LiveData<VideoEntity> mVideoById;
+    private LiveData<List<VideoEntity>> mAllVideosByCategory;
+    // mutable live data can be changed by ui controllers through setter
+    private final MutableLiveData<String> mQuery = new MutableLiveData<>();
+    private final MutableLiveData<Long> mVideoId = new MutableLiveData<>();
+    private final MutableLiveData<String> mVideoCategory = new MutableLiveData<>();
+    private DatabaseHelper mDatabaseHelper;
+
+    public VideosViewModel(Application application) {
+        super(application);
+
+        mDatabaseHelper = DatabaseHelper.getInstance();
+
+        // create database in main thread
+        mDatabaseHelper.createDb(this.getApplication());
```

Comment:
ditto. this duplicate code is bad, let the database helper handle it (so it should get the context) and return something that can initialize itself so that the caller does not need to worry about it

Reply:
yes, thank you for pointing it out, the database creation logic is handled by the database helper now.